### PR TITLE
Fix #7168: Add Brave News Widget for topics clustering

### LINKED ISSUE #7168: Brave News widget
Related ticket https://github.com/brave/roadmap/issues/582

### PR DESCRIPTION
**Pull request by kylehickinson**
## Summary of Changes

This pull request fixes #7168 

## Submitter Checklist:

- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`
- [x] New or updated UI has been tested across:
  - [x] Light & dark mode
  - [x] Different size classes (iPhone, landscape, iPad)
  - [x] Different dynamic type sizes

## Test Plan:

- Add Top News widgets on the home screen:
  - Small/single article: 15m rotation per article, refreshes after 60m
  - Medium/up to 2 articles per rotation: 15m rotation per set, refreshes after 60m
  - Large/up to 5 articles per rotation: 20m rotation per set, refreshes after 60m
- Add Top News widgets on the lock screen:
  - Rectangular/single article: 15m rotation per article, refreshes after 60m

Verify: 
- Rotations are correct, new articles appear
- Images on articles that have them are displayed
- News articles load in Brave when tapping on them

## Reviewer Checklist:

- [ ] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `bug` / `enhancement`
- [ ] Necessary [security reviews](https://github.com/brave/security/issues/new/choose) have taken place.
- [ ] Adequate unit test coverage exists to prevent regressions.
- [ ] Adequate test plan exists for QA to validate (if applicable).
- [ ] Issue and pull request is assigned to a milestone (should happen at merge time).

### CODE REVIEW

File: App/BraveWidgets/News Topics/NewsTopicsModel.swift [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+  /// - note: For now this implementation does not cache the results of the file, since the widget is not
+  /// reloaded by the app in any way yet. Hopefully we can move this code to the BraveNews module in the
+  /// future without bloating the size of the widget extension and then get all the cacheing mechanisms
+  /// from there.
+  static var live: Self {
+    .init(
+      fetchNewsTopics: {
+        do {
+          // At the moment there is only english US topics
+          let url = URL(string: "https://brave-today-cdn.brave.com/news-topic-clustering/widget_topic_news.en_US.json")!
```
Comment (Brandon-T) [inline App/BraveWidgets/News Topics/NewsTopicsModel.swift:29]: Will this URL always be the same and for en_US only? If it's possible that it won't be ASCII, we should use the `NSURL(idnString:)`

File: App/BraveWidgets/News Topics/NewsTopicsModel.swift [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+          let session = URLSession(configuration: .default, delegate: nil, delegateQueue: {
+            let queue = OperationQueue()
+            queue.maxConcurrentOperationCount = 3
+            return queue
+          }())
+          for topic in topics {
+            guard let imageURL = topic.imageURL else { continue }
+            group.addTask {
+              do {
+                let request = URLRequest(url: imageURL, timeoutInterval: 10)
```
Comment (Brandon-T) [inline App/BraveWidgets/News Topics/NewsTopicsModel.swift:64]: Is this timeout too low for slow networks?

**Review by Brandon-T — CHANGES_REQUESTED** [review]

File: App/BraveWidgets/News Topics/NewsTopicsModel.swift (hunk 1 above)
Comment (kylehickinson) [reply App/BraveWidgets/News Topics/NewsTopicsModel.swift:29]: This URL will always be this structure, the only thing that may change in the future is that the `en_US` locale + region may switch to language/region specifics for the user. No worries that it won't ever be ascii!

File: App/BraveWidgets/News Topics/NewsTopicsModel.swift (hunk 2 above)
Comment (kylehickinson) [reply App/BraveWidgets/News Topics/NewsTopicsModel.swift:64]: Its a good question, at most we're downloading 15 images at 3 at a time, so 50s on the large lists to time out every image. We dont want to have the timeout that long or the widget takes longer to show. We can maybe bump this up a bit but Im not sure what the highest we'd want to go is.

**Review by Brandon-T — APPROVED** [review]

File: App/BraveWidgets/TopNewsListWidget.swift [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+              .font(.system(size: 12))
+              .imageScale(.large)
+              .foregroundColor(Color(.braveOrange))
+            Text("Brave News")
+              .foregroundColor(Color(.braveLabel))
+              .font(.system(size: 14, weight: .bold, design: .rounded))
+          }
+          Spacer()
+          Link(destination: URL(string: "\(BraveUX.appURLScheme)://shortcut?path=\(WidgetShortcut.braveNews.rawValue)")!) {
+            Text("Read More")
```
Comment (iccub) [inline App/BraveWidgets/TopNewsListWidget.swift]: Do we translate widget text?

File: App/BraveWidgets/TopNewsWidget.swift [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+private struct TopNewsWidgetProvider: TimelineProvider {
+  private let model: NewsTopicsModel = .live
+  
+  private func thumbnailSize(for context: Context) -> CGSize {
+    let size = context.displaySize
+    let scale = context.environmentVariants.displayScale?.max() ?? 1.0
+    return .init(width: size.width * scale, height: size.height * scale)
+  }
+  
+  func getSnapshot(in context: Context, completion: @escaping (TopNewsEntry) -> Void) {
```
Comment (iccub) [inline App/BraveWidgets/TopNewsWidget.swift:69]: TimelineProvider methods are not async yet or you just prefer it completion handler based?

File: App/BraveWidgets/TopNewsWidget.swift [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+                )
+              )
+          } else {
+            LinearGradient(braveGradient: .darkGradient01)
+          }
+        }
+      )
+      .widgetURL(topic.url)
+    } else {
+      Text("Failed to load")
```
Comment (iccub) [inline App/BraveWidgets/TopNewsWidget.swift]: Do we translate it, and is this error view good enough?

File: App/BraveWidgets/TopNewsListWidget.swift (hunk 3 above)
Comment (kylehickinson) [reply App/BraveWidgets/TopNewsListWidget.swift]: No sorry they should be translated, just missed this, will fix this up along with the mock data

File: App/BraveWidgets/TopNewsWidget.swift (hunk 4 above)
Comment (kylehickinson) [reply App/BraveWidgets/TopNewsWidget.swift:69]: They are not async :(

File: App/BraveWidgets/TopNewsWidget.swift (hunk 5 above)
Comment (kylehickinson) [reply App/BraveWidgets/TopNewsWidget.swift]: We probably need a translation and maybe something nicer here

**Review by iccub — CHANGES_REQUESTED** [review]
changing from 'comment' to 'request changes' per code review feedback

File: App/BraveWidgets/TopNewsListWidget.swift [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+          .font(.system(size: 12))
+          .imageScale(.large)
+          .foregroundColor(Color(.braveOrange))
+        Text("Brave News")
+          .foregroundColor(Color(.braveLabel))
+          .font(.system(size: 14, weight: .bold, design: .rounded))
+      }
+      Spacer()
+      Link(destination: URL(string: "\(BraveUX.appURLScheme)://shortcut?path=\(WidgetShortcut.braveNews.rawValue)")!) {
+        Text("Read More")
```
Comment (iccub) [inline App/BraveWidgets/TopNewsListWidget.swift]: Not all views have been translated yet?

File: App/BraveWidgets/TopNewsListWidget.swift [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+  var entry: TopNewsListEntry
+  
+  private var headerView: some View {
+    HStack {
+      HStack(spacing: 4) {
+        Image(braveSystemName: "brave.logo")
+          .font(.system(size: 12))
+          .imageScale(.large)
+          .foregroundColor(Color(.braveOrange))
+        Text("Brave News")
```
Comment (iccub) [inline App/BraveWidgets/TopNewsListWidget.swift]: This will be most likely DNT, not sure if we want to export it regardless or not

**Review by iccub — CHANGES_REQUESTED** [review]
Localized strings missing for one view, rest looks good

File: App/BraveWidgets/TopNewsListWidget.swift (hunk 6 above)
Comment (kylehickinson) [reply App/BraveWidgets/TopNewsListWidget.swift]: Damn, missed twice! Fixed in [0781648](https://github.com/brave/brave-ios/pull/7169/commits/0781648a93366bb5dec09306d581eb0a5635df3f)

**Review by iccub — APPROVED** [review]
☁️ ☻

**Pull request merged**